File: typescript/server/src/backplane.ts
/**
 * The cross-pod backplane seam.
 *
 * In a multi-pod deployment, a connection on pod A may need events fanned in from
 * a turn processed on pod B (server-initiated pushes, shared sessions). The Rust
 * server abstracts this as a backplane with `attach(connId, sink)` /
 * `detach(connId)` (in-memory single-process, or Redis/NATS cross-pod). The
 * dispatcher/runner here are single-pod, but the connection lifecycle still runs a
 * `detach` after its read loop exits — so the seam is wired for a real backplane to
 * drop into later.
 *
 * The MVP ships an in-memory no-op backplane; the interface is what matters.
 */
import type { Frame } from './protocol.js';

/** A sink the backplane can push fanned-in events into for a given connection. */
export type BackplaneSink = (frame: Frame) => void;

/**
 * A delivery target: one connection, or every connection for a session / user / org /
 * agent. Ports the Rust reference's `Target`.
 */
export interface Target {
    kind: 'connection' | 'session' | 'user' | 'org' | 'agent';
    id: string;
}

/** The five kinds a publish may name. Anything else is the caller's error. */
export const TARGET_KINDS: readonly Target['kind'][] = ['connection', 'session', 'user', 'org', 'agent'];

/**
 * Map key for a target. NUL-separated rather than `:` because an id can legitimately
 * contain a colon (an org name, an email), which would let two distinct targets collide.
 */
function targetKey(target: Target): string {
    return `${target.kind}\u0000${target.id}`;
}

export interface Backplane {
    /** Register a connection so events addressed to it can be delivered. */
    attach(connId: string, sink: BackplaneSink): Promise<void>;
    /** Deregister a connection. ALWAYS run after a connection's read loop exits. */
    detach(connId: string): Promise<void>;
    /**
     * Associate a connection with a target, so a publish to that target reaches it.
     * Idempotent; learned over the connection's life (user/org from auth at connect,
     * session/agent as sessions resolve).
     *
     * OPTIONAL alongside {@link publish} for the same back-compat reason.
     */
    associate?(connId: string, target: Target): void;
    /**
     * Fan a frame out to every connection associated with `target`, returning how many
     * sinks it reached. The count is what `POST /admin/publish` reports as `delivered`,
     * so it must never claim a delivery that did not happen.
     *
     * OPTIONAL so a third-party backplane that predates this stays valid; the
     * admin route answers 501 when a backplane does not implement it.
     */
    publish?(target: Target, frame: Frame): number;
}

/**
 * In-memory single-process backplane: connection sinks plus a target → connections index,
 * so all five target kinds resolve locally (the Rust reference's fan-out). No cross-pod
 * fan-in — that is the Redis/NATS seam.
 */
export class InMemoryBackplane implements Backplane {
    private readonly sinks = new Map<string, BackplaneSink>();
    /** target key → the conn ids associated with it, for publish fan-out. */
    private readonly byTarget = new Map<string, Set<string>>();
    /** conn id → its target keys, so detach tears every association down. */
    private readonly byConn = new Map<string, Set<string>>();

    async attach(connId: string, sink: BackplaneSink): Promise<void> {
        this.sinks.set(connId, sink);
        // Always reachable by its own connection id.
        this.associate(connId, { kind: 'connection', id: connId });
    }

    async detach(connId: string): Promise<void> {
        this.sinks.delete(connId);
        for (const key of this.byConn.get(connId) ?? []) {
            const conns = this.byTarget.get(key);
            if (!conns) continue;
            conns.delete(connId);
            // A leaked association would resolve to a dead socket and inflate `delivered`.
            if (conns.size === 0) this.byTarget.delete(key);
        }
        this.byConn.delete(connId);
    }

    associate(connId: string, target: Target): void {
        const key = targetKey(target);
        let conns = this.byTarget.get(key);
        if (!conns) this.byTarget.set(key, (conns = new Set()));
        conns.add(connId);

        let keys = this.byConn.get(connId);
        if (!keys) this.byConn.set(connId, (keys = new Set()));
        keys.add(key);
    }

    /** Deliver to every connection for `target`. Returns the number of sinks reached. */
    publish(target: Target, frame: Frame): number {
        let delivered = 0;
        for (const connId of this.byTarget.get(targetKey(target)) ?? []) {
            const sink = this.sinks.get(connId);
            if (!sink) continue;
            sink(frame);
            delivered++;
        }
        return delivered;
    }

    /** Whether a connection is currently attached (used by tests to assert detach ran). */
    has(connId: string): boolean {
        return this.sinks.has(connId);
    }

    /** Number of currently-attached connections (used by tests to assert detach ran). */
    get size(): number {
        return this.sinks.size;
    }
}

File: go/server/backplane.go
package server

import (
	"context"
	"sync"
)

// Target is a delivery target: one connection, or every connection for a session /
// user / org / agent. A comparable struct, so it is a map key by value. Ports the
// Rust reference's Target.
type Target struct {
	Kind string
	ID   string
}

// TargetKinds are the five kinds a publish may name. Anything else is the caller's
// error.
var TargetKinds = []string{"connection", "session", "user", "org", "agent"}

// Backplane is the connection-registry seam: every connection attaches its outbound
// sink under a connection id, associates the targets it belongs to, and detaches when
// its read loop exits. Publishing to a target reaches every connection for it — the
// plug point for non-AI publishers (job status, ingestion progress, notifications)
// that need a connected client without going through an agent turn. The Go analog of
// the Rust Backplane trait, including its 5-target fan-out. The bundled
// InMemoryBackplane is single-process; Redis/NATS impls satisfy the same interface and
// additionally reach other pods' sockets.
type Backplane interface {
	// Attach registers a connection's outbound sink. sink delivers an already-built
	// event frame to the connection's writer. The connection is always reachable as
	// Target{"connection", connID}.
	Attach(ctx context.Context, connID string, sink func(event map[string]any))
	// Associate links a connection to a target so a publish to it lands here.
	// Idempotent — the session chokepoint runs on every sessionId-bearing frame, so a
	// re-association must not double-count a delivery.
	Associate(ctx context.Context, connID string, target Target)
	// Publish fans an event out to every connection associated with target, returning
	// how many sinks it reached. The count is what `POST /admin/publish` reports as
	// `delivered`, so it must never claim a delivery that did not happen.
	Publish(ctx context.Context, target Target, event map[string]any) int
	// Detach removes a connection's sink and every association to it. Always run on
	// connection teardown: a leaked association resolves to a dead socket and would
	// inflate `delivered` forever.
	Detach(ctx context.Context, connID string)
}

// InMemoryBackplane is a single-process Backplane: connection sinks plus a
// target→connections index, so all five target kinds resolve locally. No cross-POD
// fan-out (that's the Redis/NATS seam). Safe for concurrent use.
type InMemoryBackplane struct {
	// ponytail: one mutex over the whole registry. Attach/detach run once per
	// connection and publish is two map lookups, so contention is not the bottleneck —
	// shard per target only if a profile ever says otherwise.
	mu    sync.Mutex
	sinks map[string]func(event map[string]any)
	// targetConns indexes target → conn ids for publish fan-out; connTargets is the
	// reverse, so Detach tears every association down without scanning all targets.
	targetConns map[Target]map[string]struct{}
	connTargets map[string]map[Target]struct{}
}

// NewInMemoryBackplane returns an empty in-memory backplane.
func NewInMemoryBackplane() *InMemoryBackplane {
	return &InMemoryBackplane{
		sinks:       map[string]func(event map[string]any){},
		targetConns: map[Target]map[string]struct{}{},
		connTargets: map[string]map[Target]struct{}{},
	}
}

// Attach registers connID's sink and makes it reachable by its own connection id.
func (b *InMemoryBackplane) Attach(_ context.Context, connID string, sink func(event map[string]any)) {
	b.mu.Lock()
	defer b.mu.Unlock()
	b.sinks[connID] = sink
	b.link(connID, Target{Kind: "connection", ID: connID})
}

// Associate links connID to target.
func (b *InMemoryBackplane) Associate(_ context.Context, connID string, target Target) {
	b.mu.Lock()
	defer b.mu.Unlock()
	b.link(connID, target)
}

// Publish delivers event to every connection associated with target.
func (b *InMemoryBackplane) Publish(_ context.Context, target Target, event map[string]any) int {
	b.mu.Lock()
	// Snapshot under the lock, invoke OUTSIDE it: a host's sink is arbitrary code, and
	// one bad one held under the registry lock would block every attach, detach and
	// publish in the process.
	sinks := make([]func(event map[string]any), 0, len(b.targetConns[target]))
	for connID := range b.targetConns[target] {
		if sink := b.sinks[connID]; sink != nil {
			sinks = append(sinks, sink)
		}
	}
	b.mu.Unlock()
	for _, sink := range sinks {
		sink(event)
	}
	return len(sinks)
}

// Detach removes connID's sink and every association to it.
func (b *InMemoryBackplane) Detach(_ context.Context, connID string) {
	b.mu.Lock()
	defer b.mu.Unlock()
	delete(b.sinks, connID)
	for target := range b.connTargets[connID] {
		conns := b.targetConns[target]
		delete(conns, connID)
		if len(conns) == 0 {
			delete(b.targetConns, target)
		}
	}
	delete(b.connTargets, connID)
}

// link records both directions of a conn↔target association. Caller holds b.mu.
func (b *InMemoryBackplane) link(connID string, target Target) {
	if b.targetConns[target] == nil {
		b.targetConns[target] = map[string]struct{}{}
	}
	b.targetConns[target][connID] = struct{}{}
	if b.connTargets[connID] == nil {
		b.connTargets[connID] = map[Target]struct{}{}
	}
	b.connTargets[connID][target] = struct{}{}
}

// IsAttached reports whether connID currently has a sink (used by tests to verify
// detach-after-loop ran).
func (b *InMemoryBackplane) IsAttached(connID string) bool {
	b.mu.Lock()
	defer b.mu.Unlock()
	_, ok := b.sinks[connID]
	return ok
}
